### PR TITLE
Implement temperature-dependant solid domain contraints in VOF

### PR DESCRIPTION
**Pull request by AmishgaAlphonius**
# Description of the problem

- In phase change simulations, the solid domain was treated as a simple fluid with a higher viscosity (viscous penalization). It worked quite well, but it wasn't the most efficient way to compute fluid dynamics quantities. In order to improve calculation time and linear system conditioning, homogeneous constraints are imposed on velocity DOFs in a defined solid domain through a temperature range. Additionally, pressure DOFs in the same solid domain that are not connected to fluid cells are also imposed to homogeneous constraints. The temperature range is specified through a parameter subsection, namely `constrain solid domain`. Previously, the implementation was done for single-fluid simulations (#1038), but simulations such as laser powder bed fusion (LPBF) meltpool ones require two fluids when modelling.

# Description of the solution

- This PR, implements the new temperature-dependent solid domain constraints for two-phase Volume of Fluid (VOF) simulations. To select cells on which the constraints are applied, we check if the filtered phase fraction at quadrature points in the cell are within a range of accepted values. This range of accepted values is defined with the `phase fraction tolerance` parameter. This parameter defines an absolute tolerance on filtered phase fraction.

# How Has This Been Tested?

- The new feature has been tested on the **Melting Cavity** example by modelling to parts of the domain with different fluids while keeping the same properties in both fluids.
- An application test has been added to test this feature : `applications_tests/lethe-fluid/heat_transfer_vof_phase_change_constrain_solid_domain.prm` (`.output`)

# Documentation

- Documentation page was updated: (`doc/source/parameters/cfd/constrain_solid_domain.rst`).

# Comments

- The structure of `constrain_solid_domain()` changed a bit. More functions were added to avoid code line repetitions in `constrain_solid_domain_vof()`.
- Small corrections were made in other files.

### CODE REVIEW

**Review by blaisb — APPROVED** [review]
I think this is good. I'm not sure I'm a fan of the name constrain_solid_domain because this is confusing with the regular solid domain which is not based on temperature, maybe there could be a way to have a more explicit name?

**Comment by AmishgaAlphonius** [comment]
> I think this is good. I'm not sure I'm a fan of the name constrain_solid_domain because this is confusing with the regular solid domain which is not based on temperature, maybe there could be a way to have a more explicit name?
> 

I agree, would `constrain_solid_domain_with_temperature` be too long? Or how about `constrain_solid_with_temperature` or  `constrain_domain_with_temperature`?

I personally think, `constrain_solid_with_temperature` is good.

**Comment by blaisb** [comment]
> > I think this is good. I'm not sure I'm a fan of the name constrain_solid_domain because this is confusing with the regular solid domain which is not based on temperature, maybe there could be a way to have a more explicit name?
> 
> I agree, would `constrain_solid_domain_with_temperature` be too long? Or how about `constrain_solid_with_temperature` or `constrain_domain_with_temperature`?
> 
> I personally think, `constrain_solid_with_temperature` is good.

constrain_solid_with_temperature is good to me

**Comment by hepap** [comment]
Or maybe constrain_stasis @AmishgaAlphonius @blaisb ?

**Comment by AmishgaAlphonius** [comment]
> Or maybe constrain_stasis @AmishgaAlphonius @blaisb ?

I like it! I'll rename the subsection that way and call the function `constrain_stasis_with_temperature`, so if we eventually have other constraining parameters it still works!

**Review by oguevremont — APPROVED** [review]

File: applications_tests/lethe-fluid/heat_transfer_vof_phase_change_constrain_solid_domain.prm [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+end
+
+#---------------------------------------------------
+# Post-processing
+#---------------------------------------------------
+
+subsection post-processing
+  set verbosity                   = verbose
+  set calculate pressure drop     = true
+  set calculate mass conservation = false
```
[the remark this answers is not in the document]
Comment (AmishgaAlphonius) [reply applications_tests/lethe-fluid/heat_transfer_vof_phase_change_constrain_solid_domain.prm:166]: It's true that adding the barycenter will also check the velocity in some way so I'll do so!

File: source/solvers/navier_stokes_base.cc [hunk 2]
```diff
@@ -1789,79 +1820,168 @@ NavierStokesBase<dim, VectorType, DofsType>::establish_solid_domain(
...
+          get_cell_temperature_values(cell,
+                                      dof_handler_ht,
+                                      temperature_solution,
+                                      local_temperature_values);
+          add_flags_and_constrain_velocity(local_dof_indices,
+                                           local_temperature_values,
+                                           stasis_constraint_struct);
+        }
+    }
+  check_and_constrain_pressure(stasis_constraint_struct, local_dof_indices);
```
[the remark this answers is not in the document]
Comment (hepap) [reply source/solvers/navier_stokes_base.cc:1853]: Then, check_and_constrain_pressure can be inlined I think!

File: source/solvers/navier_stokes_base.cc [hunk 3]
```diff
@@ -1789,79 +1820,168 @@ NavierStokesBase<dim, VectorType, DofsType>::establish_solid_domain(
...
+          cell_is_solid = true;
+          break;
+        }
+    }
+
+  // Flag non-solid cell DOFs to identify which pressure DOFs should not
+  // be constrained.
+  if (!cell_is_solid)
+    flag_dofs_connected_to_fluid(
+      local_dof_indices, stasis_constraint_struct.dofs_are_connected_to_fluid);
```
Comment (hepap) [inline source/solvers/navier_stokes_base.cc:1963]: Check here if the if condition for solid flag is ok. In my head, if 1 quadrature point is solid than the cell should not be flag solid.

**Review by hepap — COMMENTED** [review]
Here are my comments :) 2 main things of thing to check, but nice work I can't wait to use it!

**Review by hepap — APPROVED** [review]

File: source/solvers/navier_stokes_base.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (hepap) [reply source/solvers/navier_stokes_base.cc:1853]: Yes of course for the inlining haha, I was dans le champ :´)

**Pull request merged**